Lib/_pyio.py

```diff
@@ -1674,22 +1674,31 @@ def readall(self):
                 except OSError:
                     pass
 
-        result = bytearray()
+        result = bytearray(bufsize)
+        bytes_read = 0
         while True:
-            if len(result) >= bufsize:
-                bufsize = len(result)
-                bufsize += max(bufsize, DEFAULT_BUFFER_SIZE)
-            n = bufsize - len(result)
+            if bytes_read >= bufsize:
+                # Parallels _io/fileio.c new_buffersize
+                if bufsize > 65536:
+                    addend = bufsize >> 3
+                else:
+                    addend = 256 + bufsize
+                if addend < DEFAULT_BUFFER_SIZE:
+                    addend = DEFAULT_BUFFER_SIZE
+                bufsize += addend
+                result.resize(bufsize)
+
+            assert bufsize - bytes_read > 0, "Should always try and read at least one byte"
             try:
-                chunk = os.read(self._fd, n)
+                n = os.readinto(self._fd, memoryview(result)[bytes_read:])
             except BlockingIOError:
-                if result:
+                if bytes_read:
                     break
                 return None
-            if not chunk: # reached the end of the file
+            if n == 0:  # Reached the end of the file
                 break
-            result += chunk
-
+            bytes_read += n
+        result.resize(bytes_read)
```

Comment:
`result = memoryview(result)[bytes_read:]` would avoid a truncation which can imply a memory copy in the worst case, no?

Reply:
the resize "shrink" in bytearray doesn't actually resize unless the buffer's "capacity" is 2x the requested size (https://github.com/python/cpython/blob/main/Objects/bytearrayobject.c#L201-L214). Just updates its internal "this is how long the bytes is" counter (which for things like full-file readall with known size, this should already be just one byte over the right size).

My plan currently is to make it so `bytes(bytearray(10))` and `bytearray(b'\0' * 10)` both don't copy (Ongoing discussion in https://discuss.python.org/t/add-zero-copy-conversion-of-bytearray-to-bytes-by-providing-bytes/79164). Having a `memoryview` would mean there's more than one reference to the bytearray, and I couldn't do / use that optimization.

Reply:
Ok, I'm fine with using `result.resize()` here.